Assignment3/PointAndGrid.py

```diff
@@ -0,0 +1,260 @@
+        numRows, numCols = self.getSize()
+        for i in range(numRows):
+            for j in range(numCols):
+                if self.grid[i][j] == element:
+                    result.append(Point(self, (i, j), element))
+        return result
+
+    def findNeighborsOfPoint(self, tuple):
+        """
+        Return list of Points that are neighbours of element with the given coordinates.
+        :param tuple: Tuple
+        :return: list(Point)
+        """
+        row = tuple[0]
+        col = tuple[1]
+        result = []
+        possible_coord = [(row-1, col-1),(row-1, col),(row, col-1),(row-1, col+1),
+                          (row+1, col-1),(row+1, col),(row, col+1),(row+1, col+1)]
+        for coord in possible_coord:
+            if self._validCoordinates(coord):
+                result.append(Point(self, coord, self.getElement(coord)))
+        return result
+    
+    def ifWordInGrid(self, word):
+        """
+        Return True if the word can be formed in grid, False otherwise.
+        :param word: str
+        :return: bool
+        """
+        if self.findString(word) is None:
+            return False
+        return True
+    
+    ############################## THE SOLUTION TO ASSIGNMENT 3 ##########################################
+
+    def findString(self, word):
+        """
+        Find the word in the grid where we can starting from one position move to any neighbors,
+        bub can't visit the same cell twice. Return the path if exists, None otherwise.
+        :param word: string
+        :return: bool
+        """
+        startPoints = self.findElementInGrid(word[0])
+        path = []
+        for startP in startPoints: # we go further only if the first letter of word is in grid
+            path = [startP]
+            path = self._findPathRecursively(startP, word, path, [])
+        if self._pathIsWord(path, word):
+            return path
+
+    def _pathIsWord(self, path, word):
+        return [p.value for p in path] == list(word)
+
+    def _findPathRecursively(self, point, word, path, visited):
+        """
+        Helper-function to find a word in a grid. Finds the longest path to form the word (may be only a part)
+        :param point: Point
+        :param word: str
+        :param path: list(Point)
+        :param visited: list(Point)
+        :return: None
+        """
+        if self._pathIsWord(path, word):
+            return path
+
+        possible_ways = self.findPossibleWays(path, point, word, visited)
+        if possible_ways == []:
+            visited.append(path.pop(-1)) # we at the same time pop the last element from path and append it to list visited.
+            if path:
+                self._findPathRecursively(path[-1], word, path, visited)
+
+        for neighbor in possible_ways:
+            if neighbor not in path and neighbor not in visited:
+                path.append(neighbor)
+                self._findPathRecursively(neighbor, word, path, visited)

```

Comment:
This `_findPathRecursively` call could find the path we look for. So here the return value should be checked. If it is `None` then the search should just continue. If it is not `None` it should be returned up the stack as we have found a path.